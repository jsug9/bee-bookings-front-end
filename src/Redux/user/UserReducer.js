/* eslint-disable no-param-reassign */
import axios from 'axios';
import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';

const signInEndpoint = 'https://bee-store.herokuapp.com/api/v1/signin';
const signUpEndpoint = 'https://bee-store.herokuapp.com/api/v1/signup';

export const signIn = createAsyncThunk('user/signIn', async (username) => {
  const userParams = { user: { username } };
  const { data } = await axios.post(signInEndpoint, userParams);
  return data;
});

export const sum = () => 'from user reducer';

export const signUp = createAsyncThunk('user/signUp', async (username) => {
  const userParams = { user: { username } };
  const { data } = await axios.post(signUpEndpoint, userParams);
  return data;
});

const initialState = {
  username: null,
  userId: null,
  isLoading: false,
};

const userSlice = createSlice({
  name: 'user',
  initialState,
  reducers: {
    logOutUser: (state) => {
      state.username = null;
      state.userId = null;
      state.createdUser = null;
    },
  },
  extraReducers: {
    [signIn.pending]: (state) => {
      state.isLoading = true;
    },
    [signIn.fulfilled]: (state, action) => {
      state.username = action.payload.username;
      state.userId = action.payload.user_id;
      state.isLoading = false;
    },
    [signIn.rejected]: (state) => {
      state.isLoading = false;
    },
    [signUp.pending]: (state) => {
      state.isLoading = true;
    },
    [signUp.fulfilled]: (state, action) => {
      state.username = action.payload.username;
      state.userId = action.payload.user_id;
      state.isLoading = false;
      state.createdUser = true;
    },
    [signUp.rejected]: (state) => {
      state.isLoading = false;
    },
  },
});

export const { logOutUser } = userSlice.actions;
export default userSlice.reducer;
