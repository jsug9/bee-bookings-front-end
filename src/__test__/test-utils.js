/* eslint-disable react/prop-types */
import { BrowserRouter as Router } from 'react-router-dom';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import setupStore from '../Redux/testStore';

function renderWithProviders(
  ui,
  {
    preloadedState,
    store = setupStore(preloadedState),
    ...renderOptions
  } = {},
) {
  function Wrapper({ children }) {
    return <Router><Provider store={store}>{children}</Provider></Router>;
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export * from '@testing-library/react';

export default renderWithProviders;
