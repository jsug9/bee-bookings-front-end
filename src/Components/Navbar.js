import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { NavLink } from 'react-router-dom';
import { toggleLogin } from '../Pages/LoginPage';
import { toggleSignup } from '../Pages/SignupPage';
import { logOutUser } from '../Redux/user/UserReducer';
import { clearReservations } from '../Redux/reservations/ReservationsReducer';
import '../Styles/navbar.scss';
import beeLogo from '../Assets/BeeLogo.png';

const Navbar = () => {
  const user = useSelector((state) => state.user);
  const [navbar, setNavbar] = React.useState('invisible');
  const dispatch = useDispatch();

  const handleToggleLogin = () => {
    toggleLogin();
    setNavbar(navbar === 'invisible' ? '' : 'invisible');
  };

  const handleToggleSignup = () => {
    toggleSignup();
    setNavbar(navbar === 'invisible' ? '' : 'invisible');
  };

  const handleHamburgerClick = () => {
    setNavbar(navbar === 'invisible' ? '' : 'invisible');
  };

  const handleClick = () => {
    dispatch(logOutUser());
    dispatch(clearReservations());
    setNavbar(navbar === 'invisible' ? '' : 'invisible');
  };

  const generateInactiveUserLinks = () => (
    <>
      <li className="links">
        <p role="presentation" onClick={handleToggleLogin}>Log in</p>
      </li>
      <li className="links">
        <p role="presentation" onClick={handleToggleSignup}>Sign Up</p>
      </li>
    </>
  );

  const generateActiveUserLinks = () => (
    <li>
      <p role="presentation" className="links" onClick={handleClick}>Sign Out</p>
    </li>
  );

  const links = [
    {
      id: 1,
      path: '/',
      text: 'Home',
    },
    {
      id: 2,
      path: '/add_reservation',
      text: 'Add Reservation',
    },
    {
      id: 3,
      path: '/reservations',
      text: 'Reservations',
    },
    {
      id: 4,
      path: '/add_bee',
      text: 'Add Bee',
    },
    {
      id: 5,
      path: '/delete_bee',
      text: 'Delete Bee',
    },
  ];

  return (
    <>
      <div className="hamburger" id="hamburger" role="presentation" onClick={handleHamburgerClick}>
        <div className="slice" />

        <div className="slice" />

        <div className="slice" />
      </div>
      <nav className={navbar}>
        <ul className="links">
          <li className="appLogoLi"><img className="appLogo" alt="" src={beeLogo} /></li>
          {links.map((link) => (
            <li key={link.id}>
              <NavLink
                onClick={handleHamburgerClick}
                to={link.path}
              >
                <p>{link.text}</p>
              </NavLink>
            </li>
          ))}
          {user.username ? generateActiveUserLinks() : generateInactiveUserLinks()}
        </ul>
      </nav>
    </>
  );
};

export default Navbar;
