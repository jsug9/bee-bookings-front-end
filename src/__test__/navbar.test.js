import { act } from 'react-dom/test-utils';
import Navbar from '../Components/Navbar';
import renderWithProviders, { screen } from './test-utils';
import { signIn } from '../Redux/user/UserReducer';
import setupStore from '../Redux/testStore';
import server from '../mswMocks/server';
import realStore from '../Redux/configureStore';

// Establish API mocking before all tests.
beforeAll(() => server.listen());
// Reset any request handlers that we may add during the tests,
// so they don't affect other tests.
afterEach(() => server.resetHandlers());
// Clean up after the tests are finished.
afterAll(() => server.close());

it('Navbar renders to the page', () => {
  const tree = renderWithProviders(
    <Navbar />,
  );
  expect(tree).toMatchSnapshot();
});

it('Checks that the navbar renders its appropriate default links', () => {
  renderWithProviders(
    <Navbar />,
  );
  expect(screen.getByText(/Home/)).toBeInTheDocument();
});

it('Tests the sign in functionality of the redux store', async () => {
  const store = realStore;
  await store.dispatch(signIn('AaronIsCool'));

  renderWithProviders(
    <Navbar />, { store },
  );

  expect(screen.getByText(/Sign Out/)).toBeInTheDocument();
});

it('Tests the sign out functionality of the login method triggered by the navbar', async () => {
  const store = setupStore();
  await store.dispatch(signIn('AaronIsCool'));

  act(() => {
    renderWithProviders(
      <Navbar />, { store },
    );
  });

  expect(screen.getByText(/Sign Out/)).toBeInTheDocument();
  const button = screen.getByText('Sign Out');

  act(() => {
    button.click();
  });

  expect(screen.getByText(/Log in/)).toBeInTheDocument();
});
