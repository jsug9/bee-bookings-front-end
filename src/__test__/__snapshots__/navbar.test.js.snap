// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar renders to the page 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="hamburger"
        id="hamburger"
        role="presentation"
      >
        <div
          class="slice"
        />
        <div
          class="slice"
        />
        <div
          class="slice"
        />
      </div>
      <nav
        class="invisible"
      >
        <ul
          class="links"
        >
          <li>
            <a
              aria-current="page"
              class="active"
              href="/"
            >
              <p>
                Home
              </p>
            </a>
          </li>
          <li>
            <a
              class=""
              href="/add_reservation"
            >
              <p>
                Add Reservation
              </p>
            </a>
          </li>
          <li>
            <a
              class=""
              href="/reservations"
            >
              <p>
                Reservations
              </p>
            </a>
          </li>
          <li>
            <a
              class=""
              href="/add_bee"
            >
              <p>
                Add Bee
              </p>
            </a>
          </li>
          <li>
            <a
              class=""
              href="/delete_bee"
            >
              <p>
                Delete Bee
              </p>
            </a>
          </li>
          <li
            class="links"
          >
            <p
              role="presentation"
            >
              Log in
            </p>
          </li>
          <li
            class="links"
          >
            <p
              role="presentation"
            >
              Sign Up
            </p>
          </li>
        </ul>
      </nav>
    </div>
  </body>,
  "container": <div>
    <div
      class="hamburger"
      id="hamburger"
      role="presentation"
    >
      <div
        class="slice"
      />
      <div
        class="slice"
      />
      <div
        class="slice"
      />
    </div>
    <nav
      class="invisible"
    >
      <ul
        class="links"
      >
        <li>
          <a
            aria-current="page"
            class="active"
            href="/"
          >
            <p>
              Home
            </p>
          </a>
        </li>
        <li>
          <a
            class=""
            href="/add_reservation"
          >
            <p>
              Add Reservation
            </p>
          </a>
        </li>
        <li>
          <a
            class=""
            href="/reservations"
          >
            <p>
              Reservations
            </p>
          </a>
        </li>
        <li>
          <a
            class=""
            href="/add_bee"
          >
            <p>
              Add Bee
            </p>
          </a>
        </li>
        <li>
          <a
            class=""
            href="/delete_bee"
          >
            <p>
              Delete Bee
            </p>
          </a>
        </li>
        <li
          class="links"
        >
          <p
            role="presentation"
          >
            Log in
          </p>
        </li>
        <li
          class="links"
        >
          <p
            role="presentation"
          >
            Sign Up
          </p>
        </li>
      </ul>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": Object {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
